### DISCUSSION TITLE
Introduce `inheritDirectives: false`

### DISCUSSION THREAD

**Discussion by renatodeleao** [vuejs/rfcs#480] [General]
Curious to check the community pulse on this, since I do not have enough knowledge of `@vue` core code to understand the implications of such an idea or even if it's feasible so bear with me, please.

---

One thing that broke a lot of my tests while migrating to `vue@3` was that I relied upon **heavily** — almost 100% — on a `v-test="'some-id'"` directive to target my elements and components. The only thing this directive does is conditionally add `data-test-id` [attribute when `!production`](https://frontstuff.io/unit-test-your-first-vuejs-component). 

With the `vue@2` single root node limitation, directives declared on the components always "found a way to fall through" into the root element. With `vue@3` that's no longer the case and directives are ignored if the multi-root node or **if root node does not render an element** (directly). For the first scenario, it makes total sense,  vue could not simply guess which element I wanted. In the second scenario, there is a nuance: I can still render a single node element, but not directly. Some of my components have `renderless/provider` wrappers as the root element and the actual element node is passed to its slot. In those scenarios the directive declared on the component would still fallthrough correctly in `vue@2` and now it doesn't.

### Example
```vue
<!-- VDialog.vue -->
<template>
  <!-- renderless accessibility provider for dialogs -->
  <a11y-dialog-renderless
    #default="{ backdropRef, dialogRef, titleRef, closeRef, focusRef }"
    @close="$emit('close')"
  >
    <div class="dialog-root">
       <div v-bind="backRef" />
       <div v-bind="dialogRef">
         <!-- you get the point -->
       </div>
     </div>
  </a11y-dialog-renderless>
</template>
```
and in context

```vue
<!-- someView.vue -->
<template>
   <div>
        Some view
       <v-dialog :open="isOpen" v-test="'support-dialog'">
           Need help?!
       </v-dialog>
  </div>
```

```js
// someView.spec.js
it('renders the support dialog', () => {
   const wrapper = mount(SomeView, { data: () => ({ isOpen: true }) })
   // no longer works
   expect(wrapper.find('[data-test-id="support-dialog"]').exists().toBe(true)
})
```

### Idea
The idea is to give a more or less similar treatment to `$attrs` and allow to forward directives declared on the component to a desired element node: 
1. using the `v-bind="$directives"` property which would be a map of all passed directives. `$directives` in this scenario would be prefixed with some internal identifiable string, the same way that listeners are prefixed with "on" `onX`. Since the `v-bind=` idea, would require tweaking the parsing of `v-bind` contents, an alternative idea be a new directive: `v-bind-directives` or `v-directives`.
    ```diff
    <!-- VDialog.vue -->
    <template>
      <!-- renderless accessibility provider for dialogs -->
      <a11y-dialog-renderless
        #default="{ backdropRef, dialogRef, titleRef, closeRef, focusRef }"
        @close="$emit('close')"
      >
    -    <div class="dialog-root">
    +    <div class="dialog-root" v-bind="$directives">
           <div v-bind="backRef" />
           <div v-bind="dialogRef">
             <!-- you get the point -->
           </div>
         </div>
      </a11y-dialog>
    </template>
    
    <script>
    export default {
    +  // if fallthrough behaviour is still a thing by that time
    +   inheritDirectives: false
    }
    </script>
    ```
2. Similar to `$attrs` that get their companion `useAttrs()` function, in a pure `composition-api` scenario we could also have: 
   - `useDirectives()` composable to return the provided directives map for binding.
   - `directives` could be destructured from `setup()` hook context, in a similar fashion
      ```js
       setup(props, {  directives }) {}
       ```

Is it too absurd? Is it just me? **Note:** I know that there are plenty of ways to work around this particular problem (manually adding a `data-test-id` and using `$attrs` and then tweak the compiler config to remove them but that's not the point. There are plenty of other examples that are not as easy to solve as this one, like the [O.G custom directive demo `v-focus`](https://vuejs.org/guide/reusability/custom-directives.html#introduction) for a component that does not render the `<input>` [as root](https://github.com/vuejs/core/issues/3914).

Anyways, just an idea! Cheers 👋

**Comment by fanckush** [comment]
I like the idea!
maybe changing the title to something like `Introduce inheritDirectives: false` could help it get some traction

**Reply by renatodeleao** [reply]
Done, can't hurt and welcome aboard!

**Comment by renatodeleao** [comment]
1 year after. If anyone else has suggestion on how to make the description better to create more "buzz" let me know. 

Note that even customising the directive to target a specific element via `binding.value`, if the component renders multi-root nodes or a provider/renderless component, vue simply ignores it [(and i know it's in the docs)](https://vuejs.org/guide/reusability/custom-directives.html#usage-on-components). This seems over-zealous from Vue, since a custom directive could bypass this limitation by using binding `value` and `instance` to find the target element

More demos: 
1. [I recreated the original `v-focus` custom directive form the docs](https://play.vuejs.org/#eNqVUn9r2zAQ/SpXMXAKsc3Y/gpuoBsZ69gvusH+mMZw7HOqVpaEJHsZxt99JzlOAhsdhRD77r2z3tO9gV0bk/UdshUrXGWF8eDQd2bNlWiNth4GsNjACI3VLSRETY4Qzd4o03k4gFk+d8InicdVpZXz0L/RVefgCgauAFrdKY/1ChY/US5hK1Qt1G4JvdI1XsLVeqIBTMPPUNJkBLPK7zOS474fhrK+lB2G1o9phMhZEw5bXIbGyBX9TkJat6NvEX2RvEUpNXzTVtYXCZGLfPJPzqnw2BpZeqQKoLh7vh6GODyORU5V7B7t92lL4uQVZ0ThjOoogep4eStIRODdYpOQIgY5zRf52RFsybwjhY3YZfdOK9pGvALOKt0aIdF+Ml6QA85W8+VwVpKBX+9iz9sOl3O/usPq4R/9e7cPPc4+W3Roe+TsiPnS7tBP8ObLR9zT+xEkd50k9iPgLTotu6Bxor3qVE2yz3hR7U0MDi3uq9vsPSo3mwpC48IinzPKz+tHrJ/kvshezoumWzzP39MifchH2CMlpMZGKPwQqhCkCZyXeEgQ9bmaiJu90Q4Xw4ky/idRF2lK8W5bVB40Rchr2FosH0B4KJwp1fra1fSB8AZpOg3VoodKli4kqzQmjadxFkGCZblFuX4f/ot8Kg5IJIL/bZAmfdzfeWjDkzpkiqrZAmchpjSc07F/BXb8A0pBa6w=), to allow targeting a specific element in a component. As soon as I turn it into a multi-node component, the directive stops working even if there's not an apparent strong reason for it (IMO of course).
2. Another demo for with[ a `vHighlight` directive](https://play.vuejs.org/#eNqdVF1r20AQ/Csb9UEOiaWG9sm1HdoSiEs/QlroQ68URVrbl5zuxN1JcTD+7x2dLNsJIYWCjX07c6vZ3Vmto/dVlTQ1R6No7HIrK0+OfV1NhZZlZaynNVme04bm1pQUgxrvINyd6ar2tAWTtI+0KcETOjfaeWou5WKp8PU0obXQRKWptediRIM/rE7pRupC6sUpNdoUfEyTaUcjStMeJOl8pnPGL/klU5VZ1pBraMGeakfa+CV43b2QKMn9KoF892ubI2kyVXMb+p04/6A4MbVXUjN0xWfVCtmULFBxAfVEG6Hx6aso3QI03B3El6yUoZ/GquIoPhZ6nHbNQ9tw8FxWKvOME9F4eTZdr8PlzWac4hSiu941wxJK1UREoIgI52XfLMRC90cUy5Z7zfMYmiJKkWOcHjwmOo28g8q5XCS3zmiMM/RPRLkpK6nYfqu8RBUiGvWdFVGGIu4/hZi3NZ/28XzJ+d0z8Vu3amMiurLs2DYsoh3mM4sxdPDF96+8wv8diAprBfYL4DWj9XWrsaN9qHUB2Qe8oHYWnIdJ/nAXK8/a9UW1QsPIAl9EMODHF0rfy32TvO1HjS4eGvi/dmJWXFnTyILtbin2obAW7w4c1U4enip4Dg9+aU8DuKkD+5FvPadrpYAJ3ZEvVpVxPFjvaZt/+PBoOMSWtItUMt1LpejGcnZH0tNwuKOMDwp4hUdltep8iLWA9aY9F+xCNpSrzDngWVUNgxJQOhgEld2gvNHcWDBw//yc4lUMxucWGKcB39PDfRrJ4jGb/EPFCPlgm8N9aX8RQXfaG9s+iKjdjpAwhcJ9YQdzmIYinm7Q4zntpw9U4Q0DSRN6LfS81nlrKLx29KwYHHemsnCI1SCdnNAJxTAEHPV4HE+8tJ1ym3Wb6eX5OWX23Xn2JbD5C9BF8Xg=) that fails if use an `IdProvider` wrapping component inside `AppInput` (commented out for demo purposes)

### Common valid criticisms

1. Move provider components outside and pass bindings as prop 
2. just bind the directive internally to desired element or pass a prop as binding value for it. Limitations:
   1. I cannot dynamically binding a directive based on a prop
      1. I would have to internally add "active" option to the binding.value do decide if the directive should effectively run or not on `updated` callback
   1. I cannot dynamically import a directive and bind it, there's no `defineAsyncDirective`
      1. My component now carries the directive "weight" even if i just use it a couple of instances.

### Final thoughts

I know that there are alternatives for every pattern listed, but a framework provides convenience, that's why we use them primarily. If this is not viable, I think we can safely close this thread.
